### ISSUE TITLE
Avoid to store duplicates in GridFs

### ISSUE THREAD

**Issue by DavidSichau** [vsivsi/meteor-file-collection#30]
As in my application I do not want to store the same binary file multiple times in the GridFs. Is there a way to check the md5 hash before uploading a file and storing it in GridFs.

The best would be to avoid uploading a file if the md5 hash exists and simply return the GridFs file then.

**Comment by riaan53** [comment]
That would be great! But..

I can see a few difficulties with that:
1. The md5 is only calculated when the whole file is uploaded (at the moment) on the server.
2. To calculate the md5 on the client before upload is possible but will be resource intensive. Will be quicker just to upload the file depending on size :)
3. md5 is not unique - although very very unlikely to find a duplicate file 

Method that you can give try:
Calculate the md5 of the first few bytes / last few / middle / random / combination of the file and compare to the one in gridfs before you upload. You can store the md5/md5s of the file segments in the file metedata and just perform a normal find to see if the file exist. Together with the name/size/random md5 checks it might be possible to accomplish this to a high enough degree. 

Just take a note about security. This type of information can be forged to view someone elses photos/movies/files etc. So it also depends on how sensitive the data/file is that you are recalling with this.

Let us know if you figure something out.

**Comment by DavidSichau** [comment]
@riaan53 

I solved the problem with client side md5 hash and an additional meteor method. For my case this is the best solution, as I have lots of hugh files (>0.5GB) which might get duplicated.

And I do not have security problems as the upload is only allowed by few trusted users. The rest can only access the files but cannot change them.

Here is the code for md5 generation, where [SparkMD5](https://github.com/satazor/SparkMD5) is used. It is really fast: around 5 sec for an 600mb file, so much faster than uploading the whole file.

This method is called on the ```fileAdded``` event of resumable.

```
computeHashes = function (resumableFile, cb) {
    var blobSlice = File.prototype.slice || File.prototype.mozSlice || File.prototype.webkitSlice,
        file = resumableFile.file,
        chunkSize = 2097152,                               // read in chunks of 2MB
        chunks = Math.ceil(file.size / chunkSize),
        currentChunk = 0,
        spark = new SparkMD5.ArrayBuffer(),
        frOnload = function(e) {
            console.log("read chunk nr", currentChunk + 1, "of", chunks);
            spark.append(e.target.result);                 // append array buffer
            currentChunk++;
            if (currentChunk < chunks) {
                loadNext();
            }
            else {
                var hash = spark.end();
                console.log("finished loading");
                console.info("computed hash", hash); // compute hash
                return cb(hash);
            }
        },
        frOnerror = function () {
            console.warn('oops, something went wrong.');
        };
    function loadNext() {
        var fileReader = new FileReader();
        fileReader.onload = frOnload;
        fileReader.onerror = frOnerror;
        var start = currentChunk * chunkSize,
            end = ((start + chunkSize) >= file.size) ? file.size : start + chunkSize;
        fileReader.readAsArrayBuffer(blobSlice.call(file, start, end));
    };
    loadNext();
};
```

I now use an additional method for checking if an file exists before uploading a file

```
Meteor.methods({
  existsFile: function(hash) ->
    entry = myData.findOne({md5: hash})
    return entry
});
```

**Comment by riaan53** [comment]
Thats awesome! Did not know its that quick.

Thanks for sharing the solution, will try it out.

On Friday, January 30, 2015, DavidSichau <notifications@github.com> wrote:

> Closed #30 <https://github.com/vsivsi/meteor-file-collection/issues/30>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vsivsi/meteor-file-collection/issues/30#event-227622670>
> .
>


-- 
Riaan Jacobs
riaan53@gmail.com

**Comment by vsivsi** [comment]
I :heart: issues that resolve themselves before I even read them!  :+1:
